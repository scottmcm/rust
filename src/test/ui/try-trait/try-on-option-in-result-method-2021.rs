// edition:2021

#![allow(dead_code, unused)]

///! This worked in 2018, but for 2021 we get to fix that.

struct Tricky;

impl<T: std::fmt::Debug> From<T> for Tricky {
    fn from(_: T) -> Tricky { Tricky }
}

fn foo() -> Result<(), Tricky> {
    None?; //~ ERROR the `?` operator can only be used in a function that returns `Result` or `Option`
    Ok(())
}

fn main() {
    assert!(matches!(foo(), Err(Tricky)));
}
