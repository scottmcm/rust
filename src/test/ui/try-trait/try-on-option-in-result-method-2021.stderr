error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `FromResidual`)
  --> $DIR/try-on-option-in-result-method-2021.rs:14:5
   |
LL | / fn foo() -> Result<(), Tricky> {
LL | |     None?;
   | |     ^^^^^ cannot use the `?` operator in a function that returns `Result<(), Tricky>`
LL | |     Ok(())
LL | | }
   | |_- this function should return `Result` or `Option` to accept `?`
   |
   = help: the trait `FromResidual<Option<!>>` is not implemented for `Result<(), Tricky>`
   = note: required by `from_residual`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
